### PR TITLE
HSEARCH-2092 Map the optimize() operation to Elasticsearch 'force merge' requests

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-2092

### CODE REVIEW

File: documentation/src/main/asciidoc/elasticsearch-integration.asciidoc [hunk 1]
```diff
@@ -464,6 +464,13 @@ ftQuery.enableFullTextFilter( "namedDriver" ).setParameter( "name", "liz" );
 
 For static filters, you can simply extend the `SimpleElasticsearchFilter` and provide an Elasticsearch filter in JSON form.
 
+=== Index optimization
+
+The optimization features documented in <<search-optimize>> are only partially implemented. That kind of optimization is rarely needed with recent versions of Lucene (on which Elasticsearch is based), but some of it is still provided for the very specific case of indexes meant to stay read-only for a long period of time:
+
+ * The automatic optimization is not implemented and most probably never will be.
+ * The manual optimization (`searchFactory.optimize()`) is implemented.
+
```
Comment (Sanne) [inline documentation/src/main/asciidoc/elasticsearch-integration.asciidoc]: Nice, well said.

File: elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchIndexWorkVisitor.java [hunk 2]
```diff
@@ -101,9 +99,17 @@ public ElasticsearchIndexWorkVisitor(String indexName, boolean refreshAfterWrite
...
 	@Override
 	public BackendRequest<?> visitOptimizeWork(OptimizeLuceneWork work, Void p) {
-		// TODO HSEARCH-2092 implement
-		LOG.warn( "Optimize work is not yet supported for Elasticsearch, ignoring it" );
-		return null;
+		/*
+		 * As of ES 2.1, the Optimize API has been renamed to ForceMerge,
+		 * but Jest still does not provide commands for the ForceMerge API as of
+		 * version 2.0.3
+		 * See https://github.com/searchbox-io/Jest/issues/292
```
Comment (Sanne) [inline elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchIndexWorkVisitor.java]: Do you know if the deprecated "Optimize" command is applies as a ForceMerge on a newer version of ES ?

File: elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchIndexWorkVisitor.java [hunk 3]
```diff
@@ -101,9 +99,17 @@ public ElasticsearchIndexWorkVisitor(String indexName, boolean refreshAfterWrite
...
+		 */
+		Optimize optimize = new Optimize.Builder()
+				.addIndex( indexName )
+				.build();
+
+		return new BackendRequest<JestResult>( optimize, work, indexName, refreshAfterWrite );
```
Comment (Sanne) [inline elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchIndexWorkVisitor.java]: That seems simple enough :-) :+1:

File: elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchIndexWorkVisitor.java (hunk 2 above)
Comment (yrodiere) [inline elasticsearch/src/main/java/org/hibernate/search/elasticsearch/impl/ElasticsearchIndexWorkVisitor.java]: Currently no, Jest really doesn't even know about the newer API.

But to Elasticsearch, it seems both APIs are the same: [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-optimize.html) say the Optimize API has been "renamed", and the descriptions are really close (almost copy/paste).

**Pull request merged**